### ISSUE TITLE
New Lesson: Ansible for Network Automation

### ISSUE THREAD

**Issue by Mierdin** [nre-learning/nrelabs-curriculum#97]
Need to add a lesson covering Ansible

**Comment by ffreitasalves** [comment]
Could you provide more information on this issue, please?

**Comment by Mierdin** [comment]
Apologies, I write that in a hurry. Updated....just a lesson idea

**Comment by shahbhoomi** [comment]
@Mierdin   
@ShrutiVPawaskar and I are working on this issue and we are planning to include the following in this lesson. Can you please take a look at our stages and give us your feedback/suggestions?  

## Part 1: Intro to Ansible  
* What is Ansible?  
* Ansible Structure  
* Ansible Terminology e.g. tasks, handlers,variables  
* Your First Playbook  

## Part 2: Ansible Network Modules (vendor agnostic)  
* cli-command  
* cli-config  
* net-interface  
* netconf-rpc  
* net-l2-interface  
* net-static-route  

## Part 3 Ansible Junos Modules  
* junos-facts  
* junos-config  
* junos-system  
* junos-user  
* junos-vlan  

## Part 4 : Ansible Juniper.junos role  
* junos-get-config  
* junos-cli  
* junos-get-table  
* junos-ping  
* junos-rpc

**Comment by Mierdin** [comment]
@shahbhoomi Sincere apologies for not responding earlier. I think we still need a general Ansible lesson though, before diving into Juniper specifics. The above feels like a bit of a commercial - I think there's a place for it, but perhaps after making sure the user has seen a bit more of the basics.

Also note that https://github.com/nre-learning/antidote/issues/152 was opened recently (I closed it in lieu of this one since this has been open for some time), so @ntwrkguru and @dgjnpr may also have thoughts here.

**Comment by mmellin** [comment]
Hello, I'm interested in contributing here.  I've been learning Ansible myself and am reading through the NRE contribution documents.  If I get a few free cycles and learn a little more about NRE lessons, I can post a draft lesson plan and get feedback.

@Mierdin Are there any Ansible specifics you'd like to see covered?

**Comment by Mierdin** [comment]
I"ve recently become aware of several parties working on this that haven't chimed in here yet. I highly recommend that if you're currently working on this, that you comment here now, to avoid duplicate work.

**Comment by manasasu** [comment]
Hi! Just wanted to check if this issue is still open and if there are multiple contributors, we can collaborate. 
This is my draft plan for ansible - 
What is Ansible and why is it used
What are control and managed nodes
Creating modules and tasks
Creating inventory with variables
Creating your first playbook using YAML for fetching and updating device configuration
Converting playbook into a role

**Comment by Mierdin** [comment]
I am quite sure that there are other parties working on this at this point. I will try to direct them here, but IMO this is something that is going to be well-served in the near future.

**Comment by ntwrkguru** [comment]
I’ve not had time to get the antidote and all set up in a local dev environment, and likely won’t anytime soon. 😢

Otherwise I’d contribute, despite the fact that I hate ansible for network management. But, people love it, so may as well include it.

**Comment by hillstreetblues** [comment]
Hi @manasasu we've been working in a team of two to build a lesson for this.  It's almost finished with the following structure.

1) Introduction
 - Briefly what/why is Ansible (+control/managed nodes)
 - Setting up your inventory
2) Tasks and Modules
 - Brief explanation
 - Using modules in ad-hoc command
 - A very basic playbook (doing the same ad-hoc command in a playbook)
3) Your first playbook
 - Originally we had built it to deploy two web server but since speaking with Mierdin swapped the focus to network devices.  Currently playing around with NAPALM for Ansible and sorting out what can be done.  On a non-vm environment we have tested get_facts and changing and configuration and comparing the difference which is what we hope to show in the lesson.  When trying to do this on virtual devices I have been having a couple issues but just need to get past the errors.  

For collaboration maybe we could get together and have a look at the current state of each of our lessons, we could compare and see what to merge together.  One thing I could do with help on the most is troubleshooting an sftp problem I have been having when Ansible connects to a vqfx.

**Comment by ntwrkguru** [comment]
@hillstreetblues When you say napalm, are you referring to using the core Ansible modules? If so, I would recommend >2.5 to use the vendor-neutral "magic" of the `network_cli` connection type.

Do you have a fork that you're working in? Maybe I can take a look also.

**Comment by mmellin** [comment]
I also originally had some plans to create some ansible lessons, but just
became way too busy with other things.  I'd still enjoy contributing if
possible. @hillstreetblues if you could share your fork, I could test it
out and/or add to it if you're interested.

**Comment by manasasu** [comment]
@hillstreetblues I have just started working on it. It would be great if you could share your fork and I could do some modifications to it or try to troubleshoot any errors you are facing.

**Comment by hillstreetblues** [comment]
@mmellin @manasasu 

Onto just writing lesson content at the moment which shouldn't take long to finish.  Once its written I will share my fork and we can propose changes.

Having issues using Napalm as an ad-hoc command where I don't have these problems when using a playbook.  Have you seen something like this before?

**Comment by manasasu** [comment]
@hillstreetblues Could you give us more details about the problem. What error does it throw up? Can you paste the ad-hoc command used?

**Comment by hillstreetblues** [comment]
Hey @manasasu I had to make sure my connection was defined as local in my playbook as I couldn't parse this in the ad-hoc command.  I have been away but I am back on this.  I will post more details on any errors I come up against for the future.

**Comment by Mierdin** [comment]
Hello all. I'm noticing this conversation has died off. This is one of the most-requested lessons, and it seems like there's a lot of interest in working on it as well, but I haven't seen this manifest in a pull request to date.

From what I've read in the above thread, there are potentially three lessons in play here, going from the more fundamental, to the more specific:

- General Ansible Learning
- Ansible for Networking
- Ansible for Junos (wherein we explore the numerous junos-specific Ansible modules)

We're working with the Red Hat folks to look at getting some of the content in their existing [workshops](https://github.com/ansible/workshops) into the NRE Labs curriculum, and I feel like their content sits firmly in the second realm - it's definitely applied to networking, but it's also very multi-vendor, which is a good fit for this curriculum.

Our plan for v1.1.0 is to land a single lesson that converts their earliest workshop materials into a few stages as a proof-of-concept, with the ultimate goal of converting all of their workshop material into a lesson. I would expect that this work is done within the next week or so.

Naturally, if anyone in this thread is days or weeks away from dropping a dope Ansible PR in this repo, please let me know - I intend to not render anyone's hard work obsolete. :smile: That said, even if not, there's still room to contribute Ansible content along the lines of the other two main areas I outlined above, and of course, improving any existing content.

Thanks so much for the thoughts and work thus far - and I hope this makes sense. If you have any questions please let me know.

**Comment by Mierdin** [comment]
Hello all - just wanted to give an update on this lesson. We have merged https://github.com/nre-learning/nrelabs-curriculum/pull/289, which means that we now have a generic, multi-vendor "Ansible for Networking" lesson. Whoo!

I'm going to close this issue, but as I mentioned before, I do still think there's room for more Ansible content. Not just additional stages to what was implemented in #289 (which is still planned, likely to happen in following releases) but other lessons that take a different approach. I opened https://github.com/nre-learning/nrelabs-curriculum/issues/298 to track the idea of a Junos-specific lesson, since I know there's a ton to talk about wrt those integrations.
